prismlauncher: support for glfw-wayland at runtime

spikespaz opened the pull request:
- Introduced hand-written wrapper to accomplish:
- Add support for checking `$XDG_CURRENT_SESSION` at runtime and using `glfw-wayland` if necessary.

Previously all wrapper args were in `qtWrapperArgs`.

This is undesireable for a few reasons:
1. Makes it hard to wrap again.
2. Makes it hard to debug.
3. The Qt wrapper should be focused on Qt.

I have introduced `makeWrapper` instead:
1. Less cryptic, less indirection.
2. Causes `wrapQtAppsHook` to produce a bash wrapper.
3. Allows wrapping further for
   environment-dependant runtime dependencies.## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

spikespaz commented:
We can drop e3e10d19c6d9d9de8a339cbc48c19ab54bab376b if it isn't wanted, but I thought it would be nice to make the actual wrapper and `wrapper.nix`'s `postBuild` simpler.

Scrumplex commented on pkgs/games/prismlauncher/default.nix:
```diff
@@ -38,8 +38,8 @@ rec {
     repo = "PrismLauncher";
-    rev = version;
-    sha256 = "sha256-RArg60S91YKp1Mt97a5JNfBEOf2cmuX4pK3VAx2WfqM=";
+    rev = "caf925cbad6cbd2f09ec2e55526b072b7958c5d5";
```
I don't think there is a need to bump to a development version. If users are interested in development versions, they should just use the flake `github:PrismLauncher/PrismLauncher#prismlauncher`

spikespaz replied:
That was an accident, I will fix.

Scrumplex commented on pkgs/games/prismlauncher/wrapper-tools.nix:
```diff
@@ -0,0 +1,65 @@
+{ makeSetupHook, writeScript, writeText, runtimeShell }:
+let
+  hookScript = writeScript "wrapper-tools-setup-hook.in" ''
+    #!@shell@
```
use `writeShellApplication` instead.

Scrumplex reviewed (changes requested):
I am not sure if we want this.

- This significantly increases the complexity of this derivation, while users can always do `pkgs.prismlauncher.override {glfw = pkgs.some-other-glfw; }`

- You could also just use the shell-based wrapper and set the flags similar to this package: https://github.com/NixOS/nixpkgs/blob/84e8b009ee98ece64d6faa9fbde9c88584d4a08e/pkgs/applications/networking/instant-messengers/discord/linux.nix#L108

- Unpatched `glfw-wayland` usually causes issues with Minecraft, which is why there is a [`glfw-wayland-minecraft`](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/libraries/glfw/3.x-wayland-minecraft.nix) package that's specifically patched to work with Minecraft

spikespaz commented:
> This significantly increases the complexity of this derivation, while users can always do `pkgs.prismlauncher.override {glfw = pkgs.some-other-glfw; }`

The check of `$XDG_CURRENT_SESSION` needs to be and *should* be done at runtime, just like `$NIXOS_OZONE_WL`. This means that for the people who switch between X11 and Wayland, PrismLauncher will find the correct `glfw` to use automatically.

More people than you think use both display servers. I am doing this because GTNH with *lwjgl3ify* and Java 17/18/19/20 doesn't start on Wayland.

> You could also just use the shell-based wrapper and set the flags [...]

Unless PrismLauncher has a flag for setting the path used by `dlopen` when it goes through `MangoHud::findLibrary`, that can't happen. `LD_LIBRARY_PATH` must be set to include a path to `glfw`'s object before the program is executed by the wrapper.

I already tried to use regular `makeWrapper`, and it did not work. The values of `--prefix`, `--suffix`, `--set`, etc. are all sanitized, and no amount of escaping would make it work. I'm sure you could find a loophole, but this is definitely better. I've been trying to do it your way the entire day, and ended up giving up and doing this in all of 20 minutes.

If you can figure out how to use `wrapProgram`'s `--suffix` (or another function from `make-wrapper.sh`) to get a literal expansion into the generated wrapper code, please please please tell me. I've been trying so hard, for so long, for so many packages.

> Unpatched `glfw-wayland` usually causes issues with Minecraft, which is why there is a [`glfw-wayland-minecraft`](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/libraries/glfw/3.x-wayland-minecraft.nix?rgh-link-date=2023-09-04T07%3A48%3A41Z) package that's specifically patched to work with Minecraft

I'll definitely update this in the derivation then, thank you for letting me know.

---

From Discord:

> the C-based wrapper is preferable, as it's quicker at runtime
> No need to load a full shell just to bootstrap a native application

You're talking about a matter of a couple dozen milliseconds. It's not worth it.

Think about how the majority of packages here use `makeWrapper` instead of `makeBinaryWrapper`.

The binary wrapper does have a useful comment explaining the arguments that created the wrapper, but that's just not good enough. Besides, it won't work because you can't truly *script* the binary wrapper, which is necessary for checking `$XDG_CURRENT_SESSION` at runtime.

getchoo committed:
prismlauncher: add withWaylandGLFW override

getchoo commented:
just to preface this: i agree with most of what scrumplex said. this adds a massive amount of complexity to the derivation - for what's in my opinion, little benefit

> The check of $XDG_CURRENT_SESSION needs to be and should be done at runtime, just like $NIXOS_OZONE_WL. This means that for the people who switch between X11 and Wayland, PrismLauncher will find the correct glfw to use automatically.

i think this is the first issue. checking `XDG_CURRENT_SESSION` shouldn't be "needed" in any way, as unlike the flags used by `NIXOS_OZONE_WL`, `glfw-wayland-minecraft` is a unofficial, community made patch that can have [game breaking issues](https://github.com/NixOS/nixpkgs/pull/243950#issuecomment-1638743704), and is far from supported by either mojang or upstream. we shouldn't be giving all wayland users a higher chance of instability by default, especially when the bundled and regular native version of glfw can run minecraft just fine under xwayland

now that's out of the way, i can see the point in wanting the ability to do this easily - options are awesome after all! however, i think this is the very wrong way to go about it. 

> If you can figure out how to use wrapProgram's --suffix (or another function from make-wrapper.sh) to get a literal expansion into the generated wrapper code, please please please tell me. I've been trying so hard, for so long, for so many packages.

it's not exactly what you said, but i've made a commit [here](https://github.com/getchoo/nixpkgs/commit/f7de14b8dd9a4bb4f2b4b69d7a36ff0bc17449d1) implementing a much simpler solution, without moving away from `wrapQtAppsHook`. this works as expected with both the nixpkgs version of prism and the one from our upstream flake

besides this approach being more streamlined, i think it's also a good compromise for the previous discussion about the binary and script wrappers. script wrappers for qt [have been known to cause fairly measurable latency](https://github.com/NixOS/nixpkgs/pull/225881) like scrumplex said, but in this case, they are really the only option (though this could change in the future, see [this](https://github.com/NixOS/nixpkgs/issues/172583) issue). this is definitely an issue, but assuming that `waylandSupport` - and thus a script wrapper - is *not* a default, i think it's an ok tradeoff.

spikespaz commented:
Any idea why `postBuild` doesn't include `waylandPreExec` unless it is used in the string? I tried using `qrWrapperArgs = ... ++ lib.optional waylandSupport [ "--run \"$waylandPreExec\"" ]` but it didn't show up in the wrapper.

I cherry-picked your commit @getchoo and agree this is better. Still think `wrapQtAppsHook` ought to only be used for Qt-related things, such as dependencies and flags, not necessarily runtime dependencies related to application functionality. But I reverted that change.

Replaced your `stdenv.isLinux || !waylandSupport` with logical implication (`->`) because that's what it's for. Also added another optional argument, `shellWrapper`, that can be used independently, along with an assert (with the intent of adding more should the need arise) that `waylandSupport` requires `shellWrapper`.

Ready for review.

getchoo commented:
> Still think wrapQtAppsHook ought to only be used for Qt-related things, such as dependencies and flags, not necessarily runtime dependencies related to application functionality. But I reverted that change.

i'm not really sure why? `wrapQtAppsHook` is only `wrapProgram` with some default arguments. the entire point of that function is to do exactly this - it's even the same one most apps implementing `NIXOS_OZONE_WL` use

> Also added another optional argument, shellWrapper

this option is a little odd to me. by itself it serves no real purpose besides slowing down prism's start up, and for anyone overriding the package or building it off of it, i think adding `makeWrapper` to `nativeBuildInputs` would be ok

getchoo commented on pkgs/games/prismlauncher/wrapper.nix:
```diff
@@ -21,71 +24,83 @@
+
 let
-  prismlauncherFinal = prismlauncher-unwrapped.override {
+  super = prismlauncher-unwrapped.override {
```
i think `final` would be a better name here, as it doesn't conflict with something in nixpkgs scope and is more representative of it being the *final* package after being overriden

spikespaz replied:
`final` is a bad name, because it is not final. In the context of the wrapper, we are evaluating what would be `final` or `self`.

`super` is the most accurate name, because it's the package we're wrapping. `packageSuper` is also another accurate name.

Which of these options do you like best?

- `packageSuper`
- `package'`
- `prismlauncher'`
- `unwrappedPackage`

Scrumplex replied:
I would argue `final` `prev` `super` and `self` are all bad choices as those are used in the context of overlays and overrides, so the meaning wouldn't translate well here.

`prismlauncher-unwrapped'` is fine, I guess

getchoo replied:
> In the context of the wrapper, we are evaluating what would be final or self.

yes for `prismlauncher`, but not for `prismlauncher-unwrapped`. the post-override version of it we use is the `final` of `prismlauncher-unwrapped` in the context of this expression - just as it would be in an overlay for example. i can see why there could be confusion and debate over this though, so maybe changing it to `prismlauncher-unwrapped'` would be good. it'd make it more apparent what exactly in nixpkgs we're referencing, while also being a pretty unopinionated name

spikespaz replied:
I vote for `prismlauncher'` because it is shorter, and mentioning that the package is "unwrapped" is unnecessary considering all three usages of the identifier are close together, so it's easy to see the `prismlauncher' = prismlauncher-unwrapped.override`. 

Actually, because of that, I still stand by `package'` because there's only one package being derived here. I isually keep the "main" package as `package'` and name the others by their full name + `'`.

For now, latest commit still uses `prismlauncherFinal`.

getchoo commented on pkgs/games/prismlauncher/wrapper.nix:
```diff
@@ -21,71 +24,83 @@
     inherit msaClientID gamemodeSupport;
   };
+
+  runtimePrograms = [
+    # xorg.xrandr needed for LWJGL [2.9.2, 3) https://github.com/LWJGL/lwjgl/issues/128
+    xorg.xrandr
+    mesa-demos # need glxinfo
+  ]
+  ++ additionalPrograms;
+
+  runtimeLibs = [
+    # xorg
+    xorg.libX11
+    xorg.libXext
+    xorg.libXcursor
+    xorg.libXrandr
+    xorg.libXxf86vm
+    # lwjgl
+    libpulseaudio
+    libGL
+    glfw
+    openal
+    stdenv.cc.cc.lib
+    # oshi
+    udev
+  ]
+  ++ lib.optional gamemodeSupport gamemode.lib
+  ++ lib.optional textToSpeechSupport flite
+  ++ additionalLibs;
```
these could be moved back to their original spot. it makes more sense for them to be scoped to `qtWrapperArgs` as that's the only place they're used - plus it results in slightly less thunks :p

spikespaz replied:
If you say so... I really hate deeply nested code. I was mildly offended when I first read the original, using `with` unnecessarily, extra concatenation, having deeply-nested `let` blocks, making the reading of `symlinkJoin`'s attrs arg unpleasant. Flat code and simple lists are for sure the way to go. Can you concede on this?

Scrumplex replied:
The original code wasn't really deeply nested? The reason for the separated "unnecessary" `with` block was to improve static analysis of the code. I guess it's arguable if `with` was needed at all.

getchoo replied:
i wouldn't call what's currently in master "deeply nested" honestly; there's only one `let; in` statement in the entire main body of the expression. the use of `with xorg;` is also not *super* pointless imo, but i can a see it being a very small qol thing.

but overall, i still think this is a very subjective cosmetic change and i don't see a reason to adopt it, especially when evaluation time is a bit better off without it. it's fine how it is in master really, but i would welcome the removal of `with xorg;` and the extra concatenation here

Scrumplex commented:
> Still think `wrapQtAppsHook` ought to only be used for Qt-related things, such as dependencies and flags, not necessarily runtime dependencies related to application functionality. But I reverted that change.

I hope you are aware that `wrapQtAppsHook` is strictly needed here as we are packaging a Qt application. Double wrapping is bad practice, so the best way to go here is using `qtWrapperArgs`

Just go and take a look at other Qt packages. They all just use `qtWrapperArgs` to add more stuff

Scrumplex commented:
I think there are some valid concerns from @getchoo further up (https://github.com/NixOS/nixpkgs/pull/253218#issuecomment-1705649731), that haven't been addressed yet.

- This would seemingly break the game for GNOME Wayland users with NVIDIA hardware https://github.com/NixOS/nixpkgs/pull/243950#issuecomment-1638743704
- Significant wrapper slowdown 

I tested this on my workstation (Ryzen 9 3900X, DDR4-3200 CL16-18-18-38, Samsung SSD 970 EVO):

With C-based wrapper:

```
1693897300.710495 execve("result/bin/prismlauncher", ["result/bin/prismlauncher"], 0x7ffe38fc2920 /* 156 vars */) = 0
1693897300.711167 execve("/nix/store/0i0vkaax1iy3jm6l0kxiwynywnnm9gbh-prismlauncher-7.2/bin/.prismlauncher-wrapped", ["result/bin/prismlauncher"], 0x18c62a0 /* 157 vars */) = 0
```
about 0.65ms

With Shell-based wrapper:

```
1693897393.170955 execve("result/bin/prismlauncher", ["result/bin/prismlauncher"], 0x7fff5ec32c40 /* 156 vars */) = 0
1693897393.175541 execve("/nix/store/df475fvx0ighrv2vsqgm6hc19nh1b4lb-prismlauncher-7.2/bin/.prismlauncher-wrapped", ["result/bin/prismlauncher"], 0x1f2d010 /* 157 vars */) = 0
```

about 4.6ms

To replicate this, you can run `strace -ttto strace.log --trace execve result/bin/prismlauncher` and measure the time between the `execve` syscalls

spikespaz commented:
@Scrumplex 

> This would seemingly break the game for GNOME Wayland users with NVIDIA hardware

That's why @getchoo wanted there to be a flag that is disabled by default. It is.

> Significant wrapper slowdown

From the context of milliseconds, sure, it takes longer to start a bash session than it does to launch a binary. Obviously. One is an interpreter, the other is a native instructions.

From the context of using a computer: who cares? 4ms difference? That's 1/4 the time it takes refresh the typical person's display.

Besides, it's opt-in.

---

This PR now includes nothing but the change for the optionals, `waylandSupport` and `shellWrapper`.

getchoo commented on pkgs/games/prismlauncher/wrapper.nix:
```diff
@@ -21,10 +25,16 @@
 , msaClientID ? null
 , gamemodeSupport ? stdenv.isLinux
 , textToSpeechSupport ? stdenv.isLinux
+, waylandSupport ? false
```
a small comment explaining what this does could be good

spikespaz replied:
Added, with a warning.

getchoo commented:
> This would seemingly break the game for GNOME Wayland users with NVIDIA hardware https://github.com/NixOS/nixpkgs/pull/243950#issuecomment-1638743704

for the record, testing this wrapper with the upstream flake actually works fine now (funny CSD and all), so i wouldn't say this is a major blocker. what i was originally trying to imply was that issues like that aren't exactly uncommon, and thus shouldn't be a default. for people who know this works with their hardware or want to find out, having this as an opt-in thing is pretty cool imo.

![image](https://github.com/NixOS/nixpkgs/assets/48872998/de7b23cd-30fc-4d90-acb6-6539e3b32b74)

> Significant wrapper slowdown

i'll reiterate what i said in a previous comment: this is an issue that can't be resolved until there's more progress in https://github.com/NixOS/nixpkgs/issues/172583, but as an opt in side effect from a somewhat niche override, i think it's acceptable. a warning with the override comment i mentioned [here](https://github.com/NixOS/nixpkgs/pull/253218#discussion_r1317158449) would also probably be a good idea

spikespaz committed:
prismlauncher: add comments explaining new args

spikespaz commented on pkgs/games/prismlauncher/wrapper.nix:
```diff
@@ -21,10 +25,16 @@
 , gamemodeSupport ? stdenv.isLinux
 , textToSpeechSupport ? stdenv.isLinux
+, waylandSupport ? false
+, shellWrapper ? waylandSupport
```
It's there to show how it's used. I think having `shellWrapper ? waylandSupport` is a nice visual.

getchoo replied:
> It isn't really a redundancy, it serves a purpose. The assertion is there for good reason: while not necessary for the wrapper to function, it is for the reader figuring out how this works.

if `shellWrapper` had a use on it's own outside of documentation i'd agree, but i think an addition to the comment for `waylandSupport` would be better suited for this while being just as effective at getting the message across

spikespaz replied:
I will not remove the assertion.

getchoo replied:
i will not be approving this pr then. i see no reason to have a user facing option that only adds runtime costs, especially when a comment - something that could provide much more information at zero cost - could be used in its place.

if you have a better reason as to why an assertion better helps in "figuring out how this works" than a comment, please let me know. in the meantime, "i will not remove the assertion" with no further elaboration or compromise will be a blocker

spikespaz replied:
I said I wouldn't remove the assertion, not that I wouldn't remove the argument.

The assertion makes it very clear that `makeWrapper` is required, and the comment explains why.

Do you think a reasonable compromise would be to put the `shellWrapper` in a let block (with comment above) and keep the assertion?

getchoo replied:
while i would still argue the benefit of having a variable over a comment (especially when evaluation time is considered), i would be fine with that compromise since it isn't really user facing anymore and we could revisit this without breaking any overrides people have

spikespaz commented:
By the way, PR is ready for re-review.

Scrumplex commented on pkgs/games/prismlauncher/wrapper.nix:
```diff
@@ -37,28 +59,38 @@ symlinkJoin {
+
   postBuild = ''
+    ${lib.optionalString waylandSupport ''
+      qtWrapperArgs+=(--run "$waylandPreExec")
```
Why is this separate from the existing `qtWrapperArgs` below?

spikespaz replied:
That is covered in previous discussion.
I strongly encourage you to read all of it.

Scrumplex commented on pkgs/games/prismlauncher/wrapper.nix:
```diff
@@ -21,10 +25,28 @@
+# if the option is enabled, a patched version of GLFW will be added to
+# `LD_LIBRARY_PATH` so that the launcher can use the correct one
+# depending on the desktop environment used.
+, waylandSupport ? false
```
```suggestion
, withWaylandGLFW ? false
```

spikespaz replied:
What if there are other features down the line that may depend on this option?

It would make more sense to have a generic option, that affects all features in this category.

If you think that there is no possibility of anything else ever depending on that flag, we can rename it. I agree that the name you provided would make the most sense, but also was thinking about the future.

Scrumplex replied:
I don't think that will be the case. And if that's ever the case, there might just be a breaking change to this package, which is totally fine.

spikespaz replied:
I think I want to call this `withAutoGLFW` instead, because `withWaylandGLFW` implies that it switches the package at build time. That's how the `with` prefix is usually used.

Scrumplex replied:
`with` usually just means that something is shipped with something. Not necessarily that it replaces something.

spikespaz replied:
Done, I think we should be close to merge.

spikespaz commented:
Final concerns have been addressed.
- `shellWrapper` and the comment has been moved to the let block.
- Assertions lowered below the let block.
- The unwrapped package override is now named `prismlauncher'` (contentions?).

getchoo commented on pkgs/games/prismlauncher/wrapper.nix:
```diff
@@ -21,44 +25,74 @@
 , gamemodeSupport ? stdenv.isLinux
 , textToSpeechSupport ? stdenv.isLinux
+
+# The flag `withWaylandGLFW` enables runtime-checking of `WAYLAND_DISPLAY`;
+# if the option is enabled, a patched version of GLFW will be added to
+# `LD_LIBRARY_PATH` so that the launcher can use the correct one
+# depending on the desktop environment used.
+, withWaylandGLFW ? false
```
```suggestion
  # The flag `withWaylandGLFW` enables runtime-checking of `WAYLAND_DISPLAY`;
  # if the option is enabled, a patched version of GLFW will be added to
  # `LD_LIBRARY_PATH` so that the launcher can use the correct one
  # depending on the desktop environment used.
, withWaylandGLFW ? false
```

minor formatting nitpick, otherwise lgtm

spikespaz replied:
This is the weirdest nitpick you've made so far and I have never seen formatting like that (indented comment lines) in any Nix code before. Since this is Nixpkgs, I'm going to follow the same style of the rest of the files in this repository.

If you're going to review like this, please include some evidence as to why your preferences are the most correct.

spikespaz commented:
Apologies to all the people mentioned. I messed up the rebase command.
